<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_131) on Mon Oct 16 20:37:49 BST 2017 -->
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<title>SiteSpecificLikelihoodSupportAnalysis</title>
<meta name="date" content="2017-10-16">
<link rel="stylesheet" type="text/css" href="../../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="SiteSpecificLikelihoodSupportAnalysis";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":42,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"],32:["t6","Deprecated Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/SiteSpecificLikelihoodSupportAnalysis.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SimpleCongruenceAnalysisAAWithBinariesPruningSimulation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/UnitPowerSimulation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html" target="_top">Frames</a></li>
<li><a href="SiteSpecificLikelihoodSupportAnalysis.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.qmul.sbcs.evolution.convergence.analyses</div>
<h2 title="Class SiteSpecificLikelihoodSupportAnalysis" class="title">Class SiteSpecificLikelihoodSupportAnalysis</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.qmul.sbcs.evolution.convergence.analyses.SiteSpecificLikelihoodSupportAnalysis</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="typeNameLabel">SiteSpecificLikelihoodSupportAnalysis</span>
extends java.lang.Object</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>r96: 30/3/2012
 A multiple congruence analysis to compare the SSLS values of an AA dataset on FOUR input trees and a de novo RAxML tree.
 
 
 Assumes the working directory has no existing RAxML runs with the same name (runID and alignment file name would be enough to screw it).
 
 NEW for this revision (0.0.1 r80):
                Data is now EXPLICITLY filtered for missing taxa (gaps) using the AlignedSequenceRepresentation.filterForMissingData(int filter, boolean useFactor) method.
                This method either removes:
                        [useFactor = true]: any sites with a proportion of gaps > (filter/100); or
                        [useFactor - false]: any sites with COUNT of gaps > filter
 
                useFactor = true is probably safer as there's no chance of setting filter < numTaxa (which would cause it to throw an exception). The only downside is that is harder to interpret (slightly)
 
                Also note that this means the PamlParameter cleandata should probably be set to cleandata=0 for most purposes.
 
 NEW for revision r119:
                Makes use of the new codon-trimming and stop-codon-removal routines in AlignedSequenceRepresentation (r116)
                Prints out a DNA phylip file with no invariant codons or stop codons. This should be the same as the AA one that is translated, and is indended for parallel/subsequent CODEML analyses.</dd>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Joe Parker</dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#SiteSpecificLikelihoodSupportAnalysis-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.lang.String-java.util.TreeSet-int-int-boolean-">SiteSpecificLikelihoodSupportAnalysis</a></span>(java.io.File&nbsp;data,
                                     java.io.File&nbsp;treefileH0,
                                     java.io.File&nbsp;treefileH1,
                                     java.io.File&nbsp;treefileH2,
                                     java.io.File&nbsp;treefileH3,
                                     java.io.File&nbsp;treefileH1labelled,
                                     java.io.File&nbsp;treefileH2labelled,
                                     java.io.File&nbsp;treefileH3labelled,
                                     java.io.File&nbsp;work,
                                     java.io.File&nbsp;binariesLocation,
                                     java.lang.String&nbsp;ID,
                                     java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList,
                                     int&nbsp;sitesToSimulate,
                                     int&nbsp;thisFilter,
                                     boolean&nbsp;filterThisByFactor)</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;&nbsp;</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#SiteSpecificLikelihoodSupportAnalysis-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.lang.String-java.util.TreeSet-java.lang.String:A-int-boolean-java.lang.String:A-java.lang.String:A-">SiteSpecificLikelihoodSupportAnalysis</a></span>(java.io.File&nbsp;dataSet2,
                                     java.io.File&nbsp;mainTreesFile,
                                     java.io.File&nbsp;constraintTreeFile,
                                     java.io.File&nbsp;labelledTreesFile,
                                     java.io.File&nbsp;randomTrees,
                                     java.io.File&nbsp;workDir2,
                                     java.io.File&nbsp;binaries,
                                     java.lang.String&nbsp;runID2,
                                     java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList2,
                                     java.lang.String[]&nbsp;modelsList2,
                                     int&nbsp;thisFilter,
                                     boolean&nbsp;doFactor,
                                     java.lang.String[]&nbsp;begins,
                                     java.lang.String[]&nbsp;ends)</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t6" class="tableTab"><span><a href="javascript:show(32);">Deprecated Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>java.io.File</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#getBinariesLocation--">getBinariesLocation</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>java.io.File</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#getInputAlignment--">getInputAlignment</a></span>()</code>
<div class="block">Returns a file path corresponding to the input sequence alignment</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/NewickTreeRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">NewickTreeRepresentation</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#getMainTrees--">getMainTrees</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.util.TreeSet&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#getTaxonNamesSet--">getTaxonNamesSet</a></span>()</code>
<div class="block">Return the taxon names set - the maximal list of all taxa which <i>may</i> be present in the data.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>java.io.File</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#getWorkDir--">getWorkDir</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#go--">go</a></span>()</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;
<div class="block"><span class="deprecationComment">- this method should not normally be used by CongruenceRunner, but run() instead.</span></div>
</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#isDoFullyUnconstrainedRAxML--">isDoFullyUnconstrainedRAxML</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#preValidate--">preValidate</a></span>()</code>
<div class="block">This is a utility method to initialise the NewickTreeRepresentations and AlignedSequenceRepresentation so they will be stored in the XML.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#run--">run</a></span>()</code>
<div class="block">-    prune input.trees
        -    prune labelled.trees
  -      prune random trees
        -    prune constraint.tre
  -    cat random trees with input
        -    RAxML -g on constraint.tre
        -    cat RAxML.tre with input.trees
  -    cat RAxML tree with input (if it exists)
        -    for each (model:models) {get sitewise lnL for all trees}</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#setBinariesLocation-java.io.File-">setBinariesLocation</a></span>(java.io.File&nbsp;binariesLocation)</code>&nbsp;</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#setDoFullyUnconstrainedRAxML-boolean-">setDoFullyUnconstrainedRAxML</a></span>(boolean&nbsp;doFullyUnconstrainedRAxML)</code>&nbsp;</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html#setWorkDir-java.io.File-">setWorkDir</a></span>(java.io.File&nbsp;workDir)</code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="SiteSpecificLikelihoodSupportAnalysis-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.lang.String-java.util.TreeSet-int-int-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SiteSpecificLikelihoodSupportAnalysis</h4>
<pre>@Deprecated
public&nbsp;SiteSpecificLikelihoodSupportAnalysis(java.io.File&nbsp;data,
                                                         java.io.File&nbsp;treefileH0,
                                                         java.io.File&nbsp;treefileH1,
                                                         java.io.File&nbsp;treefileH2,
                                                         java.io.File&nbsp;treefileH3,
                                                         java.io.File&nbsp;treefileH1labelled,
                                                         java.io.File&nbsp;treefileH2labelled,
                                                         java.io.File&nbsp;treefileH3labelled,
                                                         java.io.File&nbsp;work,
                                                         java.io.File&nbsp;binariesLocation,
                                                         java.lang.String&nbsp;ID,
                                                         java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList,
                                                         int&nbsp;sitesToSimulate,
                                                         int&nbsp;thisFilter,
                                                         boolean&nbsp;filterThisByFactor)</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>data</code> - - absolute file location of sequence file in fasta / phylip / nexus format.</dd>
<dd><code>treefileH0</code> - - absolute file location of H0 (null hypothesis) treefile, in Newick format.</dd>
<dd><code>treefileH1</code> - - absolute file location of H1 (alternative hypothesis 1) treefile, in Newick format.</dd>
<dd><code>treefileH2</code> - - absolute file location of H2 (alternative hypothesis 2) treefile, in Newick format.</dd>
<dd><code>treefileH3</code> - - absolute file location of H3 (alternative hypothesis 3) treefile, in Newick format.</dd>
<dd><code>work</code> - - absolute location of working directory.</dd>
<dd><code>binariesLocation</code> - - absolute location of binaries directory.</dd>
<dd><code>ID</code> - - ID string (mainly for RAxML runs).</dd>
<dd><code>taxaList</code> - - an ArrayList<String> of taxa present.</dd>
<dd><code>sitesToSimulate</code> - - integer number of sites to simulate the null distribution.</dd>
<dd><code>thisFilter</code> - - filter out sites with this many (or greater) taxa having gaps (missing data)</dd>
<dd><code>filterThisByFactor</code> - - whether to filter by % or absolute number.</dd>
</dl>
</li>
</ul>
<a name="SiteSpecificLikelihoodSupportAnalysis-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.io.File-java.lang.String-java.util.TreeSet-java.lang.String:A-int-boolean-java.lang.String:A-java.lang.String:A-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>SiteSpecificLikelihoodSupportAnalysis</h4>
<pre>public&nbsp;SiteSpecificLikelihoodSupportAnalysis(java.io.File&nbsp;dataSet2,
                                             java.io.File&nbsp;mainTreesFile,
                                             java.io.File&nbsp;constraintTreeFile,
                                             java.io.File&nbsp;labelledTreesFile,
                                             java.io.File&nbsp;randomTrees,
                                             java.io.File&nbsp;workDir2,
                                             java.io.File&nbsp;binaries,
                                             java.lang.String&nbsp;runID2,
                                             java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList2,
                                             java.lang.String[]&nbsp;modelsList2,
                                             int&nbsp;thisFilter,
                                             boolean&nbsp;doFactor,
                                             java.lang.String[]&nbsp;begins,
                                             java.lang.String[]&nbsp;ends)</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="preValidate--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>preValidate</h4>
<pre>public&nbsp;void&nbsp;preValidate()</pre>
<div class="block">This is a utility method to initialise the NewickTreeRepresentations and AlignedSequenceRepresentation so they will be stored in the XML. 
 It isn't strictly necessary.
 It <b>may</b> conflict with the same calls made later...</div>
</li>
</ul>
<a name="run--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>run</h4>
<pre>public&nbsp;void&nbsp;run()</pre>
<div class="block">-    prune input.trees
        -    prune labelled.trees
  -      prune random trees
        -    prune constraint.tre
  -    cat random trees with input
        -    RAxML -g on constraint.tre
        -    cat RAxML.tre with input.trees
  -    cat RAxML tree with input (if it exists)
        -    for each (model:models) {get sitewise lnL for all trees}</div>
</li>
</ul>
<a name="go--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>go</h4>
<pre>@Deprecated
public&nbsp;void&nbsp;go()</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;<span class="deprecationComment">- this method should not normally be used by CongruenceRunner, but run() instead.</span></div>
</li>
</ul>
<a name="isDoFullyUnconstrainedRAxML--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isDoFullyUnconstrainedRAxML</h4>
<pre>public&nbsp;boolean&nbsp;isDoFullyUnconstrainedRAxML()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the doFullyUnconstrainedRAxML</dd>
</dl>
</li>
</ul>
<a name="setDoFullyUnconstrainedRAxML-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setDoFullyUnconstrainedRAxML</h4>
<pre>public&nbsp;void&nbsp;setDoFullyUnconstrainedRAxML(boolean&nbsp;doFullyUnconstrainedRAxML)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>doFullyUnconstrainedRAxML</code> - the doFullyUnconstrainedRAxML to set</dd>
</dl>
</li>
</ul>
<a name="getInputAlignment--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getInputAlignment</h4>
<pre>public&nbsp;java.io.File&nbsp;getInputAlignment()</pre>
<div class="block">Returns a file path corresponding to the input sequence alignment</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>File - sequence alignment path</dd>
</dl>
</li>
</ul>
<a name="getWorkDir--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getWorkDir</h4>
<pre>public&nbsp;java.io.File&nbsp;getWorkDir()</pre>
</li>
</ul>
<a name="getBinariesLocation--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBinariesLocation</h4>
<pre>public&nbsp;java.io.File&nbsp;getBinariesLocation()</pre>
</li>
</ul>
<a name="getMainTrees--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMainTrees</h4>
<pre>public&nbsp;<a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/NewickTreeRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">NewickTreeRepresentation</a>&nbsp;getMainTrees()</pre>
</li>
</ul>
<a name="setWorkDir-java.io.File-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setWorkDir</h4>
<pre>public&nbsp;void&nbsp;setWorkDir(java.io.File&nbsp;workDir)</pre>
</li>
</ul>
<a name="setBinariesLocation-java.io.File-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setBinariesLocation</h4>
<pre>public&nbsp;void&nbsp;setBinariesLocation(java.io.File&nbsp;binariesLocation)</pre>
</li>
</ul>
<a name="getTaxonNamesSet--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getTaxonNamesSet</h4>
<pre>public&nbsp;java.util.TreeSet&lt;java.lang.String&gt;&nbsp;getTaxonNamesSet()</pre>
<div class="block">Return the taxon names set - the maximal list of all taxa which <i>may</i> be present in the data.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/SiteSpecificLikelihoodSupportAnalysis.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/SimpleCongruenceAnalysisAAWithBinariesPruningSimulation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/UnitPowerSimulation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/analyses/SiteSpecificLikelihoodSupportAnalysis.html" target="_top">Frames</a></li>
<li><a href="SiteSpecificLikelihoodSupportAnalysis.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
