<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_65) on Mon Aug 24 15:11:39 BST 2015 -->
<TITLE>
AlignedSequenceRepresentation
</TITLE>

<META NAME="date" CONTENT="2015-08-24">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="AlignedSequenceRepresentation";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlignedSequenceRepresentation.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/CONTEXTVersion.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlignedSequenceRepresentation.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
uk.ac.qmul.sbcs.evolution.convergence</FONT>
<BR>
Class AlignedSequenceRepresentation</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../../resources/inherit.gif" ALT="extended by "><B>uk.ac.qmul.sbcs.evolution.convergence.AlignedSequenceRepresentation</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>AlignedSequenceRepresentation</B><DT>extends java.lang.Object<DT>implements java.io.Serializable</DL>
</PRE>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>09/19/2011</DD>
<DT><B>Version:</B></DT>
  <DD>0.1</DD>
<DT><B>Author:</B></DT>
  <DD>Joe Parker, <a href="http://code.google.com/a/eclipselabs.org/u/joeparkerandthemegahairymen/">Kitson Consulting Ltd / Queen Mary, University of London.</a></DD>
<DT><B>See Also:</B><DD><A HREF="../../../../../../serialized-form.html#uk.ac.qmul.sbcs.evolution.convergence.AlignedSequenceRepresentation">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.io.File</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#file">file</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#AlignedSequenceRepresentation()">AlignedSequenceRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is the default, and currently preferred constructor for an AlignedSequenceRepresentation from a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#AlignedSequenceRepresentation(java.io.File)">AlignedSequenceRepresentation</A></B>(java.io.File&nbsp;inputFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#calculateAlignmentStats(boolean)">calculateAlignmentStats</A></B>(boolean&nbsp;evaluateGaps)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A utility method to populate basic stats (mean sitewise entropy; mean taxonwise longest ungapped sequence) for reporting.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#cloneDeep()">cloneDeep</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputFileDatatype()">determineInputFileDatatype</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputSequenceType()">determineInputSequenceType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInvariantSites()">determineInvariantSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#engapStopCodonsInDNA()">engapStopCodonsInDNA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#filterForMissingData(int, boolean)">filterForMissingData</A></B>(int&nbsp;filter,
                     boolean&nbsp;filterByFactor)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.util.HashSet&lt;java.lang.String&gt;[]&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getAminoAcidsAsFitchStates()">getAminoAcidsAsFitchStates</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A method to return a HashMap<String,HashSet<String>[]> of taxon names and sequences as states for a Fitch parsimony reconstruction by uk.ac.qmul.sbcs.evolution.convergence.TreeNode</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getFullSitesLnL(java.util.TreeMap)">getFullSitesLnL</A></B>(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.Integer,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getInputIndices()">getInputIndices</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getMeanSitewiseEntropy()">getMeanSitewiseEntropy</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getMeanTaxonwiseLongestUngappedSequence()">getMeanTaxonwiseLongestUngappedSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfInvariantSites()">getNumberOfInvariantSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfSites()">getNumberOfSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfTaxa()">getNumberOfTaxa</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.ArrayList&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getRawInput()">getRawInput</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;char[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getSequenceChars(int)">getSequenceChars</A></B>(int&nbsp;sequenceIndex)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getSitewiseEntropies(boolean)">getSitewiseEntropies</A></B>(boolean&nbsp;evaluateGaps)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Getter method (acts as accessor to private calculateSitewiseEntropies() method)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeSet&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTaxaList()">getTaxaList</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTaxon(int)">getTaxon</A></B>(int&nbsp;taxonIndex)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTransposedSites()">getTransposedSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTruncatedNamesHash()">getTruncatedNamesHash</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#hasWholeNumberOfCodons()">hasWholeNumberOfCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHash()">initialiseExpandedTranslationHash</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHashWithNoGapsInCodons()">initialiseExpandedTranslationHashWithNoGapsInCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions, <b>except</b> any with gap character)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHashWithOnlyGapsInCodons()">initialiseExpandedTranslationHashWithOnlyGapsInCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initialises the 'expanded' translation hash (includes <b>only</b> ambiguity characters with unique expansions <b>and</b> at least one gap character)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseTranslationHash()">initialiseTranslationHash</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#isAA()">isAA</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Whether this alignment has been translated or not</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#loadSequences(java.util.ArrayList, boolean)">loadSequences</A></B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;inputData,
              boolean&nbsp;reportInputRead)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#loadSequences(java.io.File, boolean)">loadSequences</A></B>(java.io.File&nbsp;inputFile,
              boolean&nbsp;reportInputRead)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printCompleteSequences()">printCompleteSequences</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A method to print the entire alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfInvariantSites()">printNumberOfInvariantSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfSites()">printNumberOfSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfTaxa()">printNumberOfTaxa</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedPrivateSubs()">printSharedPrivateSubs</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedPrivateSubs(java.lang.String, java.lang.String)">printSharedPrivateSubs</A></B>(java.lang.String&nbsp;taxon_1,
                       java.lang.String&nbsp;taxon_2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Prints substitutions shared between two taxa exclusively (no other taxa) to stdout</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedSubs()">printSharedSubs</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printShortSequences(int)">printShortSequences</A></B>(int&nbsp;printLimit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A method to print a truncated version of the alignment, including the truncatedNamesHash (with UIDs)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readFastaFile()">readFastaFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readNexusFile()">readNexusFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readPhylipFile()">readPhylipFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readXMLFile()">readXMLFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeStopCodons()">removeStopCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeTaxa(java.lang.String[])">removeTaxa</A></B>(java.lang.String[]&nbsp;taxaToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeUnambiguousGaps()">removeUnambiguousGaps</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#renameTaxon(java.lang.String, java.lang.String)">renameTaxon</A></B>(java.lang.String&nbsp;oldName,
            java.lang.String&nbsp;newName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#simulateConvergenceInInvariantSites(java.lang.String, java.lang.String[], int)">simulateConvergenceInInvariantSites</A></B>(java.lang.String&nbsp;masterTaxon,
                                    java.lang.String[]&nbsp;taxaToForce,
                                    int&nbsp;numberOfSitesToConverge)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:        constrain a subset of taxa to have the same AA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#simulateConvergenceInVariantSites(java.lang.String, java.lang.String[], int)">simulateConvergenceInVariantSites</A></B>(java.lang.String&nbsp;masterTaxon,
                                  java.lang.String[]&nbsp;taxaToForce,
                                  int&nbsp;numberOfSitesToConverge)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:        constrain a subset of taxa to have the same AA</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#stripGapsPreserveCodons()">stripGapsPreserveCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#translate(boolean)">translate</A></B>(boolean&nbsp;suppressErrors)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#trimToWholeNumberOfCodons()">trimToWholeNumberOfCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeFastaFile()">writeFastaFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeNexusFile()">writeNexusFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.io.File, boolean)">writePhylipFile</A></B>(java.io.File&nbsp;fullyPathQualifiedFileName,
                boolean&nbsp;useOriginalTaxonNames)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.lang.String)">writePhylipFile</A></B>(java.lang.String&nbsp;fullyPathQualifiedFileName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.lang.String, boolean)">writePhylipFile</A></B>(java.lang.String&nbsp;fullyPathQualifiedFileName,
                boolean&nbsp;useOriginalTaxonNames)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeXMLFile()">writeXMLFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="file"><!-- --></A><H3>
file</H3>
<PRE>
public java.io.File <B>file</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="AlignedSequenceRepresentation()"><!-- --></A><H3>
AlignedSequenceRepresentation</H3>
<PRE>
public <B>AlignedSequenceRepresentation</B>()</PRE>
<DL>
<DD>This is the default, and currently preferred constructor for an AlignedSequenceRepresentation from a file.
 <p>TODO In future a stable AlignedSequenceRepresentation(File file) constructor should be built and tested.
<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>v0.0.1</DD>
</DL>
</DL>
<HR>

<A NAME="AlignedSequenceRepresentation(java.io.File)"><!-- --></A><H3>
AlignedSequenceRepresentation</H3>
<PRE>
<FONT SIZE="-1">@Deprecated
</FONT>public <B>AlignedSequenceRepresentation</B>(java.io.File&nbsp;inputFile)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="cloneDeep()"><!-- --></A><H3>
cloneDeep</H3>
<PRE>
<FONT SIZE="-1">@Deprecated
</FONT>public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</A> <B>cloneDeep</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="loadSequences(java.io.File, boolean)"><!-- --></A><H3>
loadSequences</H3>
<PRE>
public void <B>loadSequences</B>(java.io.File&nbsp;inputFile,
                          boolean&nbsp;reportInputRead)
                   throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="determineInputFileDatatype()"><!-- --></A><H3>
determineInputFileDatatype</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</A> <B>determineInputFileDatatype</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineInputSequenceType()"><!-- --></A><H3>
determineInputSequenceType</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</A> <B>determineInputSequenceType</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="readPhylipFile()"><!-- --></A><H3>
readPhylipFile</H3>
<PRE>
public void <B>readPhylipFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>31/10/2011
 
 Added methods to parse lines that include spaces within sequence lines.
 NOTE that this assumes rawInput has NO blank lines (before data, at least).</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaFile()"><!-- --></A><H3>
readFastaFile</H3>
<PRE>
public void <B>readFastaFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>31/10/2011
 RESOLVED: 31/10/2011: TODO testing revealing some very strange behaviour in some fasta reads. work out why.
 A number of errors - 
                lastname and lastdata initialised wrongly to null not ""
                taxon names added as substr(1) not substr(1,length-1) so not now truncated
                (incidental) loadSequences() now takes boolean arg for line reporting
                MAJOR error:    lastname was always being reinitialised after sequenceHash.add() to "", so half of the time
                                                e.g. some taxa names are ignored, and their data entered in the other taxa, doubling line length
                MAJOR error:    the final taxon was never being added to sequenceHash as its match condition was not being filled.</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="readNexusFile()"><!-- --></A><H3>
readNexusFile</H3>
<PRE>
public void <B>readNexusFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="readXMLFile()"><!-- --></A><H3>
readXMLFile</H3>
<PRE>
public void <B>readXMLFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.lang.String)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.lang.String&nbsp;fullyPathQualifiedFileName)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.io.File, boolean)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.io.File&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fullyPathQualifiedFileName</CODE> - - name of the whole<DD><CODE>useOriginalTaxonNames</CODE> - - or the truncated sequence names hash (first 6 chars of original name + underscore + 3-digit padded UID)</DL>
</DD>
</DL>
<HR>

<A NAME="writeFastaFile()"><!-- --></A><H3>
writeFastaFile</H3>
<PRE>
public void <B>writeFastaFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeNexusFile()"><!-- --></A><H3>
writeNexusFile</H3>
<PRE>
public void <B>writeNexusFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeXMLFile()"><!-- --></A><H3>
writeXMLFile</H3>
<PRE>
public void <B>writeXMLFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="simulateConvergenceInVariantSites(java.lang.String, java.lang.String[], int)"><!-- --></A><H3>
simulateConvergenceInVariantSites</H3>
<PRE>
public void <B>simulateConvergenceInVariantSites</B>(java.lang.String&nbsp;masterTaxon,
                                              java.lang.String[]&nbsp;taxaToForce,
                                              int&nbsp;numberOfSitesToConverge)
                                       throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A>,
                                              <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></PRE>
<DL>
<DD>Purpose:        constrain a subset of taxa to have the same AA
        <p>execution:   
 <pre>for i in num sites{
        random indices
        get target master taxon as char array /// NB: codon triplets
        read array[indices]
        replace
}</pre>
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>masterTaxon</CODE> - - The taxon that is being treated as the master. Sites will be converged to this taxon's sequence randomly<DD><CODE>taxaToForce</CODE> - - The taxa that in which convergence is to be simulated.<DD><CODE>numberOfSitesToConverge</CODE> - - The number of sites to be forced to converge.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A></CODE> - - In cases where the taxaToForce are not present in the alignment (check case of taxa?)
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="initialiseTranslationHash()"><!-- --></A><H3>
initialiseTranslationHash</H3>
<PRE>
public void <B>initialiseTranslationHash</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>21/08/2012</DD>
<DT><B>See Also:</B><DD><CODE>translate(), removeStopCodonsInDNA()
 Initialise the translation hash (if it isn't instantiated already)</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="initialiseExpandedTranslationHash()"><!-- --></A><H3>
initialiseExpandedTranslationHash</H3>
<PRE>
public void <B>initialiseExpandedTranslationHash</B>()</PRE>
<DL>
<DD>Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions)
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="initialiseExpandedTranslationHashWithNoGapsInCodons()"><!-- --></A><H3>
initialiseExpandedTranslationHashWithNoGapsInCodons</H3>
<PRE>
public void <B>initialiseExpandedTranslationHashWithNoGapsInCodons</B>()</PRE>
<DL>
<DD>Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions, <b>except</b> any with gap character)
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="initialiseExpandedTranslationHashWithOnlyGapsInCodons()"><!-- --></A><H3>
initialiseExpandedTranslationHashWithOnlyGapsInCodons</H3>
<PRE>
public void <B>initialiseExpandedTranslationHashWithOnlyGapsInCodons</B>()</PRE>
<DL>
<DD>Initialises the 'expanded' translation hash (includes <b>only</b> ambiguity characters with unique expansions <b>and</b> at least one gap character)
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="translate(boolean)"><!-- --></A><H3>
translate</H3>
<PRE>
public void <B>translate</B>(boolean&nbsp;suppressErrors)
               throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>suppressErrors</CODE> - - controls verbose stdout commentary (TRUE = less verbose)
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></CODE><DT><B>Since:</B></DT>
  <DD>21/05/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeTaxa(java.lang.String[])"><!-- --></A><H3>
removeTaxa</H3>
<PRE>
public void <B>removeTaxa</B>(java.lang.String[]&nbsp;taxaToRemove)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>06/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeUnambiguousGaps()"><!-- --></A><H3>
removeUnambiguousGaps</H3>
<PRE>
public void <B>removeUnambiguousGaps</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>06/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeStopCodons()"><!-- --></A><H3>
removeStopCodons</H3>
<PRE>
public void <B>removeStopCodons</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>22/08/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineInvariantSites()"><!-- --></A><H3>
determineInvariantSites</H3>
<PRE>
public void <B>determineInvariantSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getRawInput()"><!-- --></A><H3>
getRawInput</H3>
<PRE>
public java.util.ArrayList&lt;java.lang.String&gt; <B>getRawInput</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfSites()"><!-- --></A><H3>
getNumberOfSites</H3>
<PRE>
public int <B>getNumberOfSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the maximum numberOfSites in the data alignment</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfTaxa()"><!-- --></A><H3>
getNumberOfTaxa</H3>
<PRE>
public int <B>getNumberOfTaxa</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the numberOfTaxa</DL>
</DD>
</DL>
<HR>

<A NAME="getTaxaList()"><!-- --></A><H3>
getTaxaList</H3>
<PRE>
public java.util.TreeSet&lt;java.lang.String&gt; <B>getTaxaList</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the taxaList (String ArrayList of untruncated taxa names.)</DL>
</DD>
</DL>
<HR>

<A NAME="getTruncatedNamesHash()"><!-- --></A><H3>
getTruncatedNamesHash</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.lang.String&gt; <B>getTruncatedNamesHash</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the truncatedNamesHash</DL>
</DD>
</DL>
<HR>

<A NAME="getTransposedSites()"><!-- --></A><H3>
getTransposedSites</H3>
<PRE>
public java.lang.String[] <B>getTransposedSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfInvariantSites()"><!-- --></A><H3>
getNumberOfInvariantSites</H3>
<PRE>
public int <B>getNumberOfInvariantSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the numberOfInvariantSites</DL>
</DD>
</DL>
<HR>

<A NAME="getMeanSitewiseEntropy()"><!-- --></A><H3>
getMeanSitewiseEntropy</H3>
<PRE>
public float <B>getMeanSitewiseEntropy</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the meanSitewiseEntropy</DL>
</DD>
</DL>
<HR>

<A NAME="getSitewiseEntropies(boolean)"><!-- --></A><H3>
getSitewiseEntropies</H3>
<PRE>
public float[] <B>getSitewiseEntropies</B>(boolean&nbsp;evaluateGaps)</PRE>
<DL>
<DD>Getter method (acts as accessor to private calculateSitewiseEntropies() method)
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>evaluateGaps</CODE> - - should gaps be treated as informative?
<DT><B>Returns:</B><DD>float[] of sitewise entropties</DL>
</DD>
</DL>
<HR>

<A NAME="getMeanTaxonwiseLongestUngappedSequence()"><!-- --></A><H3>
getMeanTaxonwiseLongestUngappedSequence</H3>
<PRE>
public float <B>getMeanTaxonwiseLongestUngappedSequence</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the meanTaxonwiseLongestUngappedSequence</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfSites()"><!-- --></A><H3>
printNumberOfSites</H3>
<PRE>
public void <B>printNumberOfSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfTaxa()"><!-- --></A><H3>
printNumberOfTaxa</H3>
<PRE>
public void <B>printNumberOfTaxa</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfInvariantSites()"><!-- --></A><H3>
printNumberOfInvariantSites</H3>
<PRE>
public void <B>printNumberOfInvariantSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printShortSequences(int)"><!-- --></A><H3>
printShortSequences</H3>
<PRE>
public void <B>printShortSequences</B>(int&nbsp;printLimit)</PRE>
<DL>
<DD>A method to print a truncated version of the alignment, including the truncatedNamesHash (with UIDs)
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>printLimit</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="printCompleteSequences()"><!-- --></A><H3>
printCompleteSequences</H3>
<PRE>
public void <B>printCompleteSequences</B>()</PRE>
<DL>
<DD>A method to print the entire alignment.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.lang.String, boolean)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.lang.String&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fullyPathQualifiedFileName</CODE> - - name of the whole<DD><CODE>useOriginalTaxonNames</CODE> - - or the truncated sequence names hash (first 6 chars of original name + underscore + 3-digit padded UID)</DL>
</DD>
</DL>
<HR>

<A NAME="renameTaxon(java.lang.String, java.lang.String)"><!-- --></A><H3>
renameTaxon</H3>
<PRE>
public boolean <B>renameTaxon</B>(java.lang.String&nbsp;oldName,
                           java.lang.String&nbsp;newName)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>oldName</CODE> - - the old name of the taxon (existing name)<DD><CODE>newName</CODE> - - the new name to change it to
<DT><B>Returns:</B><DD>true if successful.
 
 This is a VERY quickly written and tested name-changing routine.
 TODO could do with a LOT more testing... 31/10/2011<DT><B>Since:</B></DT>
  <DD>31/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getFullSitesLnL(java.util.TreeMap)"><!-- --></A><H3>
getFullSitesLnL</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</A> <B>getFullSitesLnL</B>(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>SitePatternsSSLS</CODE> - - a TreeMap<String, Float> of site patterns and their lnL values
<DT><B>Returns:</B><DD>fullSSLS - the sitewise lnL for each site in the alignment.<DT><B>Since:</B></DT>
  <DD>2011/11/30</DD>
<DT><B>See Also:</B><DD><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats"><CODE>DataSeries</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</A> <B>clone</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>clone</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="filterForMissingData(int, boolean)"><!-- --></A><H3>
filterForMissingData</H3>
<PRE>
public void <B>filterForMissingData</B>(int&nbsp;filter,
                                 boolean&nbsp;filterByFactor)
                          throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/FilterOutOfAllowableRangeException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">FilterOutOfAllowableRangeException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filter</CODE> - - either the number of missing taxa at any given site, or the proportion (0-100 out of 100) of taxa that are allowed to have missing data at any given site [e.g. if the ratio (gapped taxa/total taxa) > (filter/100), for any given site]<DD><CODE>filterByFactor</CODE> - - false to just use number of taxa missing as the threshold, true to use proportions.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/FilterOutOfAllowableRangeException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">FilterOutOfAllowableRangeException</A></CODE> - The allowable filter bounds are:
        0<filter100    - when using filterFactor = true;
  0<numberOfTaxa        - when using filterFactor = false (e.g., threshold no. of gaps per site.<DT><B>Since:</B></DT>
  <DD>0.0.1 r63/4</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getSequenceChars(int)"><!-- --></A><H3>
getSequenceChars</H3>
<PRE>
public char[] <B>getSequenceChars</B>(int&nbsp;sequenceIndex)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>-</CODE> - sequenceIndex - which sequence to get (not guaranteed sort order)
<DT><B>Returns:</B><DD>- a char array of sequence<DT><B>Since:</B></DT>
  <DD>- 09/08/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTaxon(int)"><!-- --></A><H3>
getTaxon</H3>
<PRE>
public java.lang.String <B>getTaxon</B>(int&nbsp;taxonIndex)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>-</CODE> - sequenceIndex - which taxon to get (not guaranteed sort order)
<DT><B>Returns:</B><DD>- the taxon name (String)<DT><B>Since:</B></DT>
  <DD>- 09/08/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="stripGapsPreserveCodons()"><!-- --></A><H3>
stripGapsPreserveCodons</H3>
<PRE>
public void <B>stripGapsPreserveCodons</B>()
                             throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>-</CODE> - null
<DT><B>Throws:</B>
<DD><CODE>-</CODE> - SequenceTypeNotSupportedException
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></CODE><DT><B>Since:</B></DT>
  <DD>- 21/08/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="hasWholeNumberOfCodons()"><!-- --></A><H3>
hasWholeNumberOfCodons</H3>
<PRE>
public boolean <B>hasWholeNumberOfCodons</B>()
                               throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>Boolean hasWholeNumberOfCodons - indicates whether no.codons/3 is an integer (whole codons) or not (partial codons)
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></CODE> - A method that checks whether a whole number of codons are present, e.g. if (this.numberOfSites/3) has a remainder (% > 0)
 This method should throw a SequenceTypeNotSupportedException if AA or UNDEFINED sequence types are used<DT><B>Since:</B></DT>
  <DD>21/08/2012</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="trimToWholeNumberOfCodons()"><!-- --></A><H3>
trimToWholeNumberOfCodons</H3>
<PRE>
public void <B>trimToWholeNumberOfCodons</B>()
                               throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></CODE><DT><B>Since:</B></DT>
  <DD>21/08/2012</DD>
<DT><B>See Also:</B><DD><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#hasWholeNumberOfCodons()"><CODE>hasWholeNumberOfCodons()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="engapStopCodonsInDNA()"><!-- --></A><H3>
engapStopCodonsInDNA</H3>
<PRE>
public void <B>engapStopCodonsInDNA</B>()
                          throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</A></CODE><DT><B>Since:</B></DT>
  <DD>21/08/2012</DD>
<DT><B>See Also:</B><DD><CODE>hasWholeNumberOfCodons(), trimToWholeNumberOfCodons, stripGapsPreserveCodons.</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getInputIndices()"><!-- --></A><H3>
getInputIndices</H3>
<PRE>
public java.util.HashMap&lt;java.lang.Integer,java.lang.Integer&gt; <B>getInputIndices</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getAminoAcidsAsFitchStates()"><!-- --></A><H3>
getAminoAcidsAsFitchStates</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.util.HashSet&lt;java.lang.String&gt;[]&gt; <B>getAminoAcidsAsFitchStates</B>()</PRE>
<DL>
<DD>A method to return a HashMap<String,HashSet<String>[]> of taxon names and sequences as states for a Fitch parsimony reconstruction by uk.ac.qmul.sbcs.evolution.convergence.TreeNode
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>Taxon names and sequences as HashSet<String>[]<DT><B>Since:</B></DT>
  <DD>2013/05/23</DD>
<DT><B>See Also:</B><DD><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TreeNode.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><CODE>TreeNode</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="calculateAlignmentStats(boolean)"><!-- --></A><H3>
calculateAlignmentStats</H3>
<PRE>
public void <B>calculateAlignmentStats</B>(boolean&nbsp;evaluateGaps)</PRE>
<DL>
<DD>A utility method to populate basic stats (mean sitewise entropy; mean taxonwise longest ungapped sequence) for reporting.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>evaluateGaps</CODE> - : are gaps to be evaluated in the entropy calculations?<DT><B>Since:</B></DT>
  <DD>r188, 2013/07/29</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isAA()"><!-- --></A><H3>
isAA</H3>
<PRE>
public boolean <B>isAA</B>()</PRE>
<DL>
<DD>Whether this alignment has been translated or not
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>- TRUE if alignment's SequenceCodingType is 'AA'<DT><B>See Also:</B><DD><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence"><CODE>SequenceCodingType</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="loadSequences(java.util.ArrayList, boolean)"><!-- --></A><H3>
loadSequences</H3>
<PRE>
public void <B>loadSequences</B>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;inputData,
                          boolean&nbsp;reportInputRead)
                   throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="printSharedSubs()"><!-- --></A><H3>
printSharedSubs</H3>
<PRE>
public void <B>printSharedSubs</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printSharedPrivateSubs()"><!-- --></A><H3>
printSharedPrivateSubs</H3>
<PRE>
public void <B>printSharedPrivateSubs</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printSharedPrivateSubs(java.lang.String, java.lang.String)"><!-- --></A><H3>
printSharedPrivateSubs</H3>
<PRE>
public void <B>printSharedPrivateSubs</B>(java.lang.String&nbsp;taxon_1,
                                   java.lang.String&nbsp;taxon_2)</PRE>
<DL>
<DD>Prints substitutions shared between two taxa exclusively (no other taxa) to stdout
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>taxon_1</CODE> - <DD><CODE>taxon_2</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="simulateConvergenceInInvariantSites(java.lang.String, java.lang.String[], int)"><!-- --></A><H3>
simulateConvergenceInInvariantSites</H3>
<PRE>
public void <B>simulateConvergenceInInvariantSites</B>(java.lang.String&nbsp;masterTaxon,
                                                java.lang.String[]&nbsp;taxaToForce,
                                                int&nbsp;numberOfSitesToConverge)
                                         throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A>,
                                                <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></PRE>
<DL>
<DD>Purpose:        constrain a subset of taxa to have the same AA
        <p>execution:   
 <pre>for i in num sites{
        random indices
        get target master taxon as char array /// NB: codon triplets
        read array[indices]
        replace
}</pre>
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>masterTaxon</CODE> - - The taxon that is being treated as the master. Sites will be converged to this taxon's sequence randomly<DD><CODE>taxaToForce</CODE> - - The taxa that in which convergence is to be simulated.<DD><CODE>numberOfSitesToConverge</CODE> - - The number of sites to be forced to converge.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A></CODE> - - In cases where the taxaToForce are not present in the alignment (check case of taxa?)
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlignedSequenceRepresentation.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/CONTEXTVersion.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlignedSequenceRepresentation.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
