<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_65) on Wed Feb 11 17:40:58 GMT 2015 -->
<TITLE>
MultiHnCongruenceAnalysis
</TITLE>

<META NAME="date" CONTENT="2015-02-11">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="MultiHnCongruenceAnalysis";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/MultiHnCongruenceAnalysis.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/CopyOfCongruenceAnalysisAAWithBinariesPruningSimulationFilteringCDFs.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis1000BinCDFs.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MultiHnCongruenceAnalysis.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
uk.ac.qmul.sbcs.evolution.convergence.analyses</FONT>
<BR>
Class MultiHnCongruenceAnalysis</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../../../resources/inherit.gif" ALT="extended by "><B>uk.ac.qmul.sbcs.evolution.convergence.analyses.MultiHnCongruenceAnalysis</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>MultiHnCongruenceAnalysis</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>r96: 30/3/2012
 A multiple congruence analysis to compare the SSLS values of an AA dataset on FOUR input trees and a de novo RAxML tree.
 
 
 Assumes the working directory has no existing RAxML runs with the same name (runID and alignment file name would be enough to screw it).
 
 NEW for this revision (0.0.1 r80):
                Data is now EXPLICITLY filtered for missing taxa (gaps) using the AlignedSequenceRepresentation.filterForMissingData(int filter, boolean useFactor) method.
                This method either removes:
                        [useFactor = true]: any sites with a proportion of gaps > (filter/100); or
                        [useFactor - false]: any sites with COUNT of gaps > filter
 
                useFactor = true is probably safer as there's no chance of setting filter < numTaxa (which would cause it to throw an exception). The only downside is that is harder to interpret (slightly)
 
                Also note that this means the PamlParameter cleandata should probably be set to cleandata=0 for most purposes.
 
 NEW for revision r119:
                Makes use of the new codon-trimming and stop-codon-removal routines in AlignedSequenceRepresentation (r116)
                Prints out a DNA phylip file with no invariant codons or stop codons. This should be the same as the AA one that is translated, and is indended for parallel/subsequent CODEML analyses.</DD>
<DT><B>Author:</B></DT>
  <DD>Joe Parker</DD>
</DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#MultiHnCongruenceAnalysis(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.lang.String, java.util.TreeSet, int, int, boolean)">MultiHnCongruenceAnalysis</A></B>(java.io.File&nbsp;data,
                          java.io.File&nbsp;treefileH0,
                          java.io.File&nbsp;treefileH1,
                          java.io.File&nbsp;treefileH2,
                          java.io.File&nbsp;treefileH3,
                          java.io.File&nbsp;treefileH1labelled,
                          java.io.File&nbsp;treefileH2labelled,
                          java.io.File&nbsp;treefileH3labelled,
                          java.io.File&nbsp;work,
                          java.io.File&nbsp;binariesLocation,
                          java.lang.String&nbsp;ID,
                          java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList,
                          int&nbsp;sitesToSimulate,
                          int&nbsp;thisFilter,
                          boolean&nbsp;filterThisByFactor)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>This constructor (and most of this class) initially cloned from one of 
 the other uk.ac.qmul.sbcs.evolution.convergence.analyses (probably 
 MultiHnCongruenceAnalysisNoCDFSitewiseLikelihoodOutputTemp). As a result,
 it takes a variety of arguments (treefileH0, treefileH1.. treefileH3 etc
 for example) pertaining to analyses with a fixed number of phylogenies (4)
 as this was the prototype specification (April 2011-Summer 2012). 
 <p>These, fixed-number-of-topologies-variables, were then analysed via
 the deprecated <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method.</p>
 <p>Both this constructor / these variables, and the 
 <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method are deprecated. Instead, a 
 MultiHnCongruenceAnalysis should be instantiated using the multi-tree 
 constructor and variables, 
 <CODE>MultiHnCongruenceAnalysis(File dataSet2, File mainTreesFile, 
 File constraintTreeFile, File labelledTreesFile, File workDir2, File binaries, 
 String runID2, TreeSet<String> taxaList2, String[] modelsList2, int thisFilter, 
 boolean doFactor)</CODE>, and analysis called with the <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#run()"><CODE>run()</CODE></A>
 method, as this allows for phylogeny sets of arbitary length.</p></I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#MultiHnCongruenceAnalysis(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.lang.String, java.util.TreeSet, java.lang.String[], int, boolean)">MultiHnCongruenceAnalysis</A></B>(java.io.File&nbsp;dataSet2,
                          java.io.File&nbsp;mainTreesFile,
                          java.io.File&nbsp;constraintTreeFile,
                          java.io.File&nbsp;labelledTreesFile,
                          java.io.File&nbsp;workDir2,
                          java.io.File&nbsp;binaries,
                          java.lang.String&nbsp;runID2,
                          java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList2,
                          java.lang.String[]&nbsp;modelsList2,
                          int&nbsp;thisFilter,
                          boolean&nbsp;doFactor)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()">go</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>- this method should not normally be used by CongruenceRunner, but run() instead. 
 <b>However NOTE THAT</b> the go() and run() methods are expected to run with two separate constructors: 
 <ul><li>The 
 <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method should be run on an object that has been
 instantiated with the <CODE>MultiHnCongruenceAnalysis.MultiHnCongruenceAnalysis(File data, 
          File treefileH0, File treefileH1, File treefileH2, File treefileH3, File treefileH1labelled, 
          File treefileH2labelled, File treefileH3labelled, File work, File binariesLocation, String ID, 
          TreeSet&lt;String&gt; taxaList, int sitesToSimulate, int thisFilter, boolean filterThisByFactor)</CODE>
 constructor - <b>this combination dates from 2012 and is DEPRECATED</b></li>
 <li>The <b>(preferred)</b> <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#run()"><CODE>run()</CODE></A> method is intended to be called
 on objects that have been created via the <b>(preferred)</b> <CODE>MultiHnCongruenceAnalysis.MultiHnCongruenceAnalysis(File dataSet2, File mainTreesFile, 
          File constraintTreeFile, File labelledTreesFile, File workDir2, File binaries, String runID2, 
          TreeSet&lt;String&gt; taxaList2, String[] modelsList2, int thisFilter, boolean doFactor)</CODE> constructor.</p></I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#run()">run</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-    prune input.trees
        -    prune labelled.trees
        -    prune constraint.tre
        -    RAxML -g on constraint.tre
        -    cat RAxML.tre with input.trees
        -    for each (model:models) {get sitewise lnL for all trees}</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="MultiHnCongruenceAnalysis(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.lang.String, java.util.TreeSet, int, int, boolean)"><!-- --></A><H3>
MultiHnCongruenceAnalysis</H3>
<PRE>
<FONT SIZE="-1">@Deprecated
</FONT>public <B>MultiHnCongruenceAnalysis</B>(java.io.File&nbsp;data,
                                            java.io.File&nbsp;treefileH0,
                                            java.io.File&nbsp;treefileH1,
                                            java.io.File&nbsp;treefileH2,
                                            java.io.File&nbsp;treefileH3,
                                            java.io.File&nbsp;treefileH1labelled,
                                            java.io.File&nbsp;treefileH2labelled,
                                            java.io.File&nbsp;treefileH3labelled,
                                            java.io.File&nbsp;work,
                                            java.io.File&nbsp;binariesLocation,
                                            java.lang.String&nbsp;ID,
                                            java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList,
                                            int&nbsp;sitesToSimulate,
                                            int&nbsp;thisFilter,
                                            boolean&nbsp;filterThisByFactor)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>This constructor (and most of this class) initially cloned from one of 
 the other uk.ac.qmul.sbcs.evolution.convergence.analyses (probably 
 MultiHnCongruenceAnalysisNoCDFSitewiseLikelihoodOutputTemp). As a result,
 it takes a variety of arguments (treefileH0, treefileH1.. treefileH3 etc
 for example) pertaining to analyses with a fixed number of phylogenies (4)
 as this was the prototype specification (April 2011-Summer 2012). 
 <p>These, fixed-number-of-topologies-variables, were then analysed via
 the deprecated <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method.</p>
 <p>Both this constructor / these variables, and the 
 <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method are deprecated. Instead, a 
 MultiHnCongruenceAnalysis should be instantiated using the multi-tree 
 constructor and variables, 
 <CODE>MultiHnCongruenceAnalysis(File dataSet2, File mainTreesFile, 
 File constraintTreeFile, File labelledTreesFile, File workDir2, File binaries, 
 String runID2, TreeSet<String> taxaList2, String[] modelsList2, int thisFilter, 
 boolean doFactor)</CODE>, and analysis called with the <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#run()"><CODE>run()</CODE></A>
 method, as this allows for phylogeny sets of arbitary length.</p></I>
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>data</CODE> - - absolute file location of sequence file in fasta / phylip / nexus format.<DD><CODE>treefileH0</CODE> - - absolute file location of H0 (null hypothesis) treefile, in Newick format.<DD><CODE>treefileH1</CODE> - - absolute file location of H1 (alternative hypothesis 1) treefile, in Newick format.<DD><CODE>treefileH2</CODE> - - absolute file location of H2 (alternative hypothesis 2) treefile, in Newick format.<DD><CODE>treefileH3</CODE> - - absolute file location of H3 (alternative hypothesis 3) treefile, in Newick format.<DD><CODE>work</CODE> - - absolute location of working directory.<DD><CODE>binariesLocation</CODE> - - absolute location of binaries directory.<DD><CODE>ID</CODE> - - ID string (mainly for RAxML runs).<DD><CODE>taxaList</CODE> - - an ArrayList<String> of taxa present.<DD><CODE>sitesToSimulate</CODE> - - integer number of sites to simulate the null distribution.<DD><CODE>thisFilter</CODE> - - filter out sites with this many (or greater) taxa having gaps (missing data)<DD><CODE>filterThisByFactor</CODE> - - whether to filter by % or absolute number.<DT><B>See Also:</B><DD><CODE>MultiHnCongruenceAnalysis(File dataSet2, File mainTreesFile, 
 File constraintTreeFile, File labelledTreesFile, File workDir2, File binaries, 
 String runID2, TreeSet<String> taxaList2, String[] modelsList2, int thisFilter, 
 boolean doFactor)}</CODE>, 
<CODE>MultiHnCongruenceAnalysis#go()}</CODE>, 
<CODE>MultiHnCongruenceAnalysis#run()}</CODE></DL>
</DL>
<HR>

<A NAME="MultiHnCongruenceAnalysis(java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.io.File, java.lang.String, java.util.TreeSet, java.lang.String[], int, boolean)"><!-- --></A><H3>
MultiHnCongruenceAnalysis</H3>
<PRE>
public <B>MultiHnCongruenceAnalysis</B>(java.io.File&nbsp;dataSet2,
                                 java.io.File&nbsp;mainTreesFile,
                                 java.io.File&nbsp;constraintTreeFile,
                                 java.io.File&nbsp;labelledTreesFile,
                                 java.io.File&nbsp;workDir2,
                                 java.io.File&nbsp;binaries,
                                 java.lang.String&nbsp;runID2,
                                 java.util.TreeSet&lt;java.lang.String&gt;&nbsp;taxaList2,
                                 java.lang.String[]&nbsp;modelsList2,
                                 int&nbsp;thisFilter,
                                 boolean&nbsp;doFactor)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="run()"><!-- --></A><H3>
run</H3>
<PRE>
public void <B>run</B>()</PRE>
<DL>
<DD>-    prune input.trees
        -    prune labelled.trees
        -    prune constraint.tre
        -    RAxML -g on constraint.tre
        -    cat RAxML.tre with input.trees
        -    for each (model:models) {get sitewise lnL for all trees}
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="go()"><!-- --></A><H3>
go</H3>
<PRE>
public void <B>go</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>- this method should not normally be used by CongruenceRunner, but run() instead. 
 <b>However NOTE THAT</b> the go() and run() methods are expected to run with two separate constructors: 
 <ul><li>The 
 <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#go()"><CODE>go()</CODE></A> method should be run on an object that has been
 instantiated with the <CODE>MultiHnCongruenceAnalysis.MultiHnCongruenceAnalysis(File data, 
          File treefileH0, File treefileH1, File treefileH2, File treefileH3, File treefileH1labelled, 
          File treefileH2labelled, File treefileH3labelled, File work, File binariesLocation, String ID, 
          TreeSet&lt;String&gt; taxaList, int sitesToSimulate, int thisFilter, boolean filterThisByFactor)</CODE>
 constructor - <b>this combination dates from 2012 and is DEPRECATED</b></li>
 <li>The <b>(preferred)</b> <A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html#run()"><CODE>run()</CODE></A> method is intended to be called
 on objects that have been created via the <b>(preferred)</b> <CODE>MultiHnCongruenceAnalysis.MultiHnCongruenceAnalysis(File dataSet2, File mainTreesFile, 
          File constraintTreeFile, File labelledTreesFile, File workDir2, File binaries, String runID2, 
          TreeSet&lt;String&gt; taxaList2, String[] modelsList2, int thisFilter, boolean doFactor)</CODE> constructor.</p></I>
<P>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/MultiHnCongruenceAnalysis.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/CopyOfCongruenceAnalysisAAWithBinariesPruningSimulationFilteringCDFs.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../../uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis1000BinCDFs.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.analyses"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/analyses/MultiHnCongruenceAnalysis.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="MultiHnCongruenceAnalysis.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
