<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_131) on Mon Oct 16 20:37:52 BST 2017 -->
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<title>AlignedSequenceRepresentation</title>
<meta name="date" content="2017-10-16">
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="AlignedSequenceRepresentation";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":42,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":42,"i9":10,"i10":10,"i11":10,"i12":10,"i13":10,"i14":10,"i15":10,"i16":10,"i17":10,"i18":10,"i19":10,"i20":10,"i21":10,"i22":10,"i23":10,"i24":10,"i25":10,"i26":10,"i27":10,"i28":10,"i29":10,"i30":10,"i31":10,"i32":10,"i33":10,"i34":10,"i35":10,"i36":10,"i37":10,"i38":10,"i39":10,"i40":10,"i41":10,"i42":10,"i43":10,"i44":10,"i45":10,"i46":10,"i47":10,"i48":10,"i49":10,"i50":10,"i51":10,"i52":10,"i53":10,"i54":10,"i55":10,"i56":10,"i57":10,"i58":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"],32:["t6","Deprecated Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AlignedSequenceRepresentation.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignmentTranslationTableType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top">Frames</a></li>
<li><a href="AlignedSequenceRepresentation.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">uk.ac.qmul.sbcs.evolution.convergence</div>
<h2 title="Class AlignedSequenceRepresentation" class="title">Class AlignedSequenceRepresentation</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>uk.ac.qmul.sbcs.evolution.convergence.AlignedSequenceRepresentation</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable</dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">AlignedSequenceRepresentation</span>
extends java.lang.Object
implements java.io.Serializable</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>09/19/2011</dd>
<dt><span class="simpleTagLabel">Version:</span></dt>
<dd>0.1</dd>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Joe Parker, <a href="http://code.google.com/a/eclipselabs.org/u/joeparkerandthemegahairymen/">Kitson Consulting Ltd / Queen Mary, University of London.</a></dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../serialized-form.html#uk.ac.qmul.sbcs.evolution.convergence.AlignedSequenceRepresentation">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.io.File</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#file">file</a></span></code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#AlignedSequenceRepresentation--">AlignedSequenceRepresentation</a></span>()</code>
<div class="block">This is the default, and currently preferred constructor for an AlignedSequenceRepresentation from a file.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#AlignedSequenceRepresentation-java.io.File-">AlignedSequenceRepresentation</a></span>(java.io.File&nbsp;inputFile)</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t6" class="tableTab"><span><a href="javascript:show(32);">Deprecated Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#calculateAlignmentStats-boolean-">calculateAlignmentStats</a></span>(boolean&nbsp;evaluateGaps)</code>
<div class="block">A utility method to populate basic stats (mean sitewise entropy; mean taxonwise longest ungapped sequence) for reporting.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#clone--">clone</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#cloneDeep--">cloneDeep</a></span>()</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputFileDatatype--">determineInputFileDatatype</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputSequenceType--">determineInputSequenceType</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInvariantSites--">determineInvariantSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#engapStopCodonsInDNA--">engapStopCodonsInDNA</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#filterForMissingData-int-boolean-">filterForMissingData</a></span>(int&nbsp;filter,
                    boolean&nbsp;filterByFactor)</code>&nbsp;</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#forceTranslationIgnoringSequenceType--">forceTranslationIgnoringSequenceType</a></span>()</code>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;
<div class="block"><span class="deprecationComment">Debug only.</span></div>
</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.String,java.util.HashSet&lt;java.lang.String&gt;[]&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getAminoAcidsAsFitchStates--">getAminoAcidsAsFitchStates</a></span>()</code>
<div class="block">A method to return a HashMap<String,HashSet<String>[]> of taxon names and sequences as states for a Fitch parsimony reconstruction by uk.ac.qmul.sbcs.evolution.convergence.TreeNode</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getFullSitesLnL-java.util.TreeMap-">getFullSitesLnL</a></span>(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</code>&nbsp;</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.Integer,java.lang.Integer&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getInputIndices--">getInputIndices</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>float</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getMeanSitewiseEntropy--">getMeanSitewiseEntropy</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>float</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getMeanTaxonwiseLongestUngappedSequence--">getMeanTaxonwiseLongestUngappedSequence</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfInvariantSites--">getNumberOfInvariantSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfSites--">getNumberOfSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfTaxa--">getNumberOfTaxa</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getRawInput--">getRawInput</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i18" class="altColor">
<td class="colFirst"><code>char[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getSequenceChars-int-">getSequenceChars</a></span>(int&nbsp;sequenceIndex)</code>&nbsp;</td>
</tr>
<tr id="i19" class="rowColor">
<td class="colFirst"><code>float[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getSitewiseEntropies-boolean-">getSitewiseEntropies</a></span>(boolean&nbsp;evaluateGaps)</code>
<div class="block">Getter method (acts as accessor to private calculateSitewiseEntropies() method)</div>
</td>
</tr>
<tr id="i20" class="altColor">
<td class="colFirst"><code>java.util.TreeSet&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTaxaList--">getTaxaList</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i21" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTaxon-int-">getTaxon</a></span>(int&nbsp;taxonIndex)</code>&nbsp;</td>
</tr>
<tr id="i22" class="altColor">
<td class="colFirst"><code>java.lang.String[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTransposedSites--">getTransposedSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i23" class="rowColor">
<td class="colFirst"><code>java.util.HashMap&lt;java.lang.String,java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTruncatedNamesHash--">getTruncatedNamesHash</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i24" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#hasWholeNumberOfCodons--">hasWholeNumberOfCodons</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i25" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHash--">initialiseExpandedTranslationHash</a></span>()</code>
<div class="block">Initialises the 'MEGA expanded' translation hash (includes all ambiguity characters with unique expansions).</div>
</td>
</tr>
<tr id="i26" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHashWithNoGapsInCodons--">initialiseExpandedTranslationHashWithNoGapsInCodons</a></span>()</code>
<div class="block">Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions, <b>except</b> any with gap character).</div>
</td>
</tr>
<tr id="i27" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseExpandedTranslationHashWithOnlyGapsInCodons--">initialiseExpandedTranslationHashWithOnlyGapsInCodons</a></span>()</code>
<div class="block">Initialises the 'expanded' translation hash (includes <b>only</b> ambiguity characters with unique expansions <b>and</b> at least one gap character).</div>
</td>
</tr>
<tr id="i28" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#initialiseTranslationHash--">initialiseTranslationHash</a></span>()</code>
<div class="block">Initialise the translation hash (if it isn't instantiated already).</div>
</td>
</tr>
<tr id="i29" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#isAA--">isAA</a></span>()</code>
<div class="block">Whether this alignment has been translated or not</div>
</td>
</tr>
<tr id="i30" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#loadSequences-java.util.ArrayList-boolean-">loadSequences</a></span>(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;inputData,
             boolean&nbsp;reportInputRead)</code>&nbsp;</td>
</tr>
<tr id="i31" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#loadSequences-java.io.File-boolean-">loadSequences</a></span>(java.io.File&nbsp;inputFile,
             boolean&nbsp;reportInputRead)</code>&nbsp;</td>
</tr>
<tr id="i32" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printCompleteSequences--">printCompleteSequences</a></span>()</code>
<div class="block">A method to print the entire alignment.</div>
</td>
</tr>
<tr id="i33" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfInvariantSites--">printNumberOfInvariantSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i34" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfSites--">printNumberOfSites</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i35" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfTaxa--">printNumberOfTaxa</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i36" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedPrivateSubs--">printSharedPrivateSubs</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i37" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedPrivateSubs-java.lang.String-java.lang.String-">printSharedPrivateSubs</a></span>(java.lang.String&nbsp;taxon_1,
                      java.lang.String&nbsp;taxon_2)</code>
<div class="block">Prints substitutions shared between two taxa exclusively (no other taxa) to stdout</div>
</td>
</tr>
<tr id="i38" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printSharedSubs--">printSharedSubs</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i39" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printShortSequences-int-">printShortSequences</a></span>(int&nbsp;printLimit)</code>
<div class="block">A method to print a truncated version of the alignment, including the truncatedNamesHash (with UIDs)</div>
</td>
</tr>
<tr id="i40" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readFastaFile--">readFastaFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i41" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readNexusFile--">readNexusFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i42" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readPhylipFile--">readPhylipFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i43" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readXMLFile--">readXMLFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i44" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeStopCodons--">removeStopCodons</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i45" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeTaxa-java.lang.String:A-">removeTaxa</a></span>(java.lang.String[]&nbsp;taxaToRemove)</code>&nbsp;</td>
</tr>
<tr id="i46" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeUnambiguousGaps--">removeUnambiguousGaps</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i47" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#renameTaxon-java.lang.String-java.lang.String-">renameTaxon</a></span>(java.lang.String&nbsp;oldName,
           java.lang.String&nbsp;newName)</code>&nbsp;</td>
</tr>
<tr id="i48" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#simulateConvergenceInInvariantSites-java.lang.String-java.lang.String:A-int-">simulateConvergenceInInvariantSites</a></span>(java.lang.String&nbsp;masterTaxon,
                                   java.lang.String[]&nbsp;taxaToForce,
                                   int&nbsp;numberOfSitesToConverge)</code>
<div class="block">Purpose:        constrain a subset of taxa to have the same AA</div>
</td>
</tr>
<tr id="i49" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#simulateConvergenceInVariantSites-java.lang.String-java.lang.String:A-int-">simulateConvergenceInVariantSites</a></span>(java.lang.String&nbsp;masterTaxon,
                                 java.lang.String[]&nbsp;taxaToForce,
                                 int&nbsp;numberOfSitesToConverge)</code>
<div class="block">Purpose:        constrain a subset of taxa to have the same AA</div>
</td>
</tr>
<tr id="i50" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#stripGapsPreserveCodons--">stripGapsPreserveCodons</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i51" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#translate-boolean-">translate</a></span>(boolean&nbsp;suppressErrors)</code>&nbsp;</td>
</tr>
<tr id="i52" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#trimToWholeNumberOfCodons--">trimToWholeNumberOfCodons</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i53" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeFastaFile--">writeFastaFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i54" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeNexusFile--">writeNexusFile</a></span>()</code>&nbsp;</td>
</tr>
<tr id="i55" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile-java.io.File-boolean-">writePhylipFile</a></span>(java.io.File&nbsp;fullyPathQualifiedFileName,
               boolean&nbsp;useOriginalTaxonNames)</code>&nbsp;</td>
</tr>
<tr id="i56" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile-java.lang.String-">writePhylipFile</a></span>(java.lang.String&nbsp;fullyPathQualifiedFileName)</code>&nbsp;</td>
</tr>
<tr id="i57" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile-java.lang.String-boolean-">writePhylipFile</a></span>(java.lang.String&nbsp;fullyPathQualifiedFileName,
               boolean&nbsp;useOriginalTaxonNames)</code>&nbsp;</td>
</tr>
<tr id="i58" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeXMLFile--">writeXMLFile</a></span>()</code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="file">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>file</h4>
<pre>public&nbsp;java.io.File file</pre>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="AlignedSequenceRepresentation--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>AlignedSequenceRepresentation</h4>
<pre>public&nbsp;AlignedSequenceRepresentation()</pre>
<div class="block">This is the default, and currently preferred constructor for an AlignedSequenceRepresentation from a file.
 <p>TODO In future a stable AlignedSequenceRepresentation(File file) constructor should be built and tested.</div>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>v0.0.1</dd>
</dl>
</li>
</ul>
<a name="AlignedSequenceRepresentation-java.io.File-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>AlignedSequenceRepresentation</h4>
<pre>@Deprecated
public&nbsp;AlignedSequenceRepresentation(java.io.File&nbsp;inputFile)</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="cloneDeep--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>cloneDeep</h4>
<pre>@Deprecated
public&nbsp;<a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</a>&nbsp;cloneDeep()</pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;</div>
</li>
</ul>
<a name="loadSequences-java.io.File-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>loadSequences</h4>
<pre>public&nbsp;boolean&nbsp;loadSequences(java.io.File&nbsp;inputFile,
                             boolean&nbsp;reportInputRead)
                      throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</a></pre>
<dl>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</a></code></dd>
</dl>
</li>
</ul>
<a name="determineInputFileDatatype--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>determineInputFileDatatype</h4>
<pre>public&nbsp;<a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</a>&nbsp;determineInputFileDatatype()</pre>
</li>
</ul>
<a name="determineInputSequenceType--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>determineInputSequenceType</h4>
<pre>public&nbsp;<a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</a>&nbsp;determineInputSequenceType()</pre>
</li>
</ul>
<a name="readPhylipFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readPhylipFile</h4>
<pre>public&nbsp;void&nbsp;readPhylipFile()</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>31/10/2011
 
 Added methods to parse lines that include spaces within sequence lines.
 NOTE that this assumes rawInput has NO blank lines (before data, at least).</dd>
</dl>
</li>
</ul>
<a name="readFastaFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readFastaFile</h4>
<pre>public&nbsp;void&nbsp;readFastaFile()</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>31/10/2011
 RESOLVED: 31/10/2011: TODO testing revealing some very strange behaviour in some fasta reads. work out why.
 A number of errors - 
                lastname and lastdata initialised wrongly to null not ""
                taxon names added as substr(1) not substr(1,length-1) so not now truncated
                (incidental) loadSequences() now takes boolean arg for line reporting
                MAJOR error:    lastname was always being reinitialised after sequenceHash.add() to "", so half of the time
                                                e.g. some taxa names are ignored, and their data entered in the other taxa, doubling line length
                MAJOR error:    the final taxon was never being added to sequenceHash as its match condition was not being filled.</dd>
</dl>
</li>
</ul>
<a name="readNexusFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readNexusFile</h4>
<pre>public&nbsp;void&nbsp;readNexusFile()</pre>
</li>
</ul>
<a name="readXMLFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readXMLFile</h4>
<pre>public&nbsp;void&nbsp;readXMLFile()</pre>
</li>
</ul>
<a name="writePhylipFile-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writePhylipFile</h4>
<pre>public&nbsp;void&nbsp;writePhylipFile(java.lang.String&nbsp;fullyPathQualifiedFileName)</pre>
</li>
</ul>
<a name="writePhylipFile-java.io.File-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writePhylipFile</h4>
<pre>public&nbsp;void&nbsp;writePhylipFile(java.io.File&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>fullyPathQualifiedFileName</code> - - name of the whole</dd>
<dd><code>useOriginalTaxonNames</code> - - or the truncated sequence names hash (first 6 chars of original name + underscore + 3-digit padded UID)</dd>
</dl>
</li>
</ul>
<a name="writeFastaFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeFastaFile</h4>
<pre>public&nbsp;void&nbsp;writeFastaFile()</pre>
</li>
</ul>
<a name="writeNexusFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeNexusFile</h4>
<pre>public&nbsp;void&nbsp;writeNexusFile()</pre>
</li>
</ul>
<a name="writeXMLFile--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeXMLFile</h4>
<pre>public&nbsp;void&nbsp;writeXMLFile()</pre>
</li>
</ul>
<a name="simulateConvergenceInVariantSites-java.lang.String-java.lang.String:A-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>simulateConvergenceInVariantSites</h4>
<pre>public&nbsp;void&nbsp;simulateConvergenceInVariantSites(java.lang.String&nbsp;masterTaxon,
                                              java.lang.String[]&nbsp;taxaToForce,
                                              int&nbsp;numberOfSitesToConverge)
                                       throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</a>,
                                              <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</a></pre>
<div class="block">Purpose:        constrain a subset of taxa to have the same AA
        <p>execution:   
 <pre>for i in num sites{
        random indices
        get target master taxon as char array /// NB: codon triplets
        read array[indices]
        replace
}</pre></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>masterTaxon</code> - - The taxon that is being treated as the master. Sites will be converged to this taxon's sequence randomly</dd>
<dd><code>taxaToForce</code> - - The taxa that in which convergence is to be simulated.</dd>
<dd><code>numberOfSitesToConverge</code> - - The number of sites to be forced to converge.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</a></code> - - In cases where the taxaToForce are not present in the alignment (check case of taxa?)</dd>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</a></code></dd>
</dl>
</li>
</ul>
<a name="initialiseTranslationHash--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialiseTranslationHash</h4>
<pre>public&nbsp;void&nbsp;initialiseTranslationHash()</pre>
<div class="block">Initialise the translation hash (if it isn't instantiated already). Any codons not listed here will be translated as a gap '-' character.</div>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>21/08/2012</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><code>translate(), removeStopCodonsInDNA()</code></dd>
</dl>
</li>
</ul>
<a name="initialiseExpandedTranslationHash--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialiseExpandedTranslationHash</h4>
<pre>public&nbsp;void&nbsp;initialiseExpandedTranslationHash()</pre>
<div class="block">Initialises the 'MEGA expanded' translation hash (includes all ambiguity characters with unique expansions). Some of these codons include gaps - this is unorthodox behaviour.
 Any codons not listed here will be translated as a gap '-' character.</div>
</li>
</ul>
<a name="initialiseExpandedTranslationHashWithNoGapsInCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialiseExpandedTranslationHashWithNoGapsInCodons</h4>
<pre>public&nbsp;void&nbsp;initialiseExpandedTranslationHashWithNoGapsInCodons()</pre>
<div class="block">Initialises the 'expanded' translation hash (includes all ambiguity characters with unique expansions, <b>except</b> any with gap character). None of these codons include gaps. Any codons containing gaps will be trainslated as a gap.</div>
</li>
</ul>
<a name="initialiseExpandedTranslationHashWithOnlyGapsInCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialiseExpandedTranslationHashWithOnlyGapsInCodons</h4>
<pre>public&nbsp;void&nbsp;initialiseExpandedTranslationHashWithOnlyGapsInCodons()</pre>
<div class="block">Initialises the 'expanded' translation hash (includes <b>only</b> ambiguity characters with unique expansions <b>and</b> at least one gap character). ALL of these codons include gaps - this is unorthodox behaviour.</div>
</li>
</ul>
<a name="translate-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>translate</h4>
<pre>public&nbsp;void&nbsp;translate(boolean&nbsp;suppressErrors)
               throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>suppressErrors</code> - - controls verbose stdout commentary (TRUE = less verbose)</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>java.lang.Exception</code></dd>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code></dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>21/05/2012</dd>
</dl>
</li>
</ul>
<a name="removeTaxa-java.lang.String:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>removeTaxa</h4>
<pre>public&nbsp;void&nbsp;removeTaxa(java.lang.String[]&nbsp;taxaToRemove)</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>06/10/2011</dd>
</dl>
</li>
</ul>
<a name="removeUnambiguousGaps--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>removeUnambiguousGaps</h4>
<pre>public&nbsp;void&nbsp;removeUnambiguousGaps()</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>06/10/2011</dd>
</dl>
</li>
</ul>
<a name="removeStopCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>removeStopCodons</h4>
<pre>public&nbsp;void&nbsp;removeStopCodons()</pre>
<dl>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>22/08/2012</dd>
</dl>
</li>
</ul>
<a name="determineInvariantSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>determineInvariantSites</h4>
<pre>public&nbsp;void&nbsp;determineInvariantSites()</pre>
</li>
</ul>
<a name="getRawInput--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getRawInput</h4>
<pre>public&nbsp;java.util.ArrayList&lt;java.lang.String&gt;&nbsp;getRawInput()</pre>
</li>
</ul>
<a name="getNumberOfSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfSites</h4>
<pre>public&nbsp;int&nbsp;getNumberOfSites()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the maximum numberOfSites in the data alignment</dd>
</dl>
</li>
</ul>
<a name="getNumberOfTaxa--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfTaxa</h4>
<pre>public&nbsp;int&nbsp;getNumberOfTaxa()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the numberOfTaxa</dd>
</dl>
</li>
</ul>
<a name="getTaxaList--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTaxaList</h4>
<pre>public&nbsp;java.util.TreeSet&lt;java.lang.String&gt;&nbsp;getTaxaList()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the taxaList (String ArrayList of untruncated taxa names.)</dd>
</dl>
</li>
</ul>
<a name="getTruncatedNamesHash--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTruncatedNamesHash</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.String&gt;&nbsp;getTruncatedNamesHash()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the truncatedNamesHash</dd>
</dl>
</li>
</ul>
<a name="getTransposedSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTransposedSites</h4>
<pre>public&nbsp;java.lang.String[]&nbsp;getTransposedSites()</pre>
</li>
</ul>
<a name="getNumberOfInvariantSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNumberOfInvariantSites</h4>
<pre>public&nbsp;int&nbsp;getNumberOfInvariantSites()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the numberOfInvariantSites</dd>
</dl>
</li>
</ul>
<a name="getMeanSitewiseEntropy--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanSitewiseEntropy</h4>
<pre>public&nbsp;float&nbsp;getMeanSitewiseEntropy()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the meanSitewiseEntropy</dd>
</dl>
</li>
</ul>
<a name="getSitewiseEntropies-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSitewiseEntropies</h4>
<pre>public&nbsp;float[]&nbsp;getSitewiseEntropies(boolean&nbsp;evaluateGaps)</pre>
<div class="block">Getter method (acts as accessor to private calculateSitewiseEntropies() method)</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>evaluateGaps</code> - - should gaps be treated as informative?</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>float[] of sitewise entropties</dd>
</dl>
</li>
</ul>
<a name="getMeanTaxonwiseLongestUngappedSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMeanTaxonwiseLongestUngappedSequence</h4>
<pre>public&nbsp;float&nbsp;getMeanTaxonwiseLongestUngappedSequence()</pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the meanTaxonwiseLongestUngappedSequence</dd>
</dl>
</li>
</ul>
<a name="printNumberOfSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printNumberOfSites</h4>
<pre>public&nbsp;void&nbsp;printNumberOfSites()</pre>
</li>
</ul>
<a name="printNumberOfTaxa--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printNumberOfTaxa</h4>
<pre>public&nbsp;void&nbsp;printNumberOfTaxa()</pre>
</li>
</ul>
<a name="printNumberOfInvariantSites--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printNumberOfInvariantSites</h4>
<pre>public&nbsp;void&nbsp;printNumberOfInvariantSites()</pre>
</li>
</ul>
<a name="printShortSequences-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printShortSequences</h4>
<pre>public&nbsp;void&nbsp;printShortSequences(int&nbsp;printLimit)</pre>
<div class="block">A method to print a truncated version of the alignment, including the truncatedNamesHash (with UIDs)</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>printLimit</code> - </dd>
</dl>
</li>
</ul>
<a name="printCompleteSequences--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printCompleteSequences</h4>
<pre>public&nbsp;void&nbsp;printCompleteSequences()</pre>
<div class="block">A method to print the entire alignment.</div>
</li>
</ul>
<a name="writePhylipFile-java.lang.String-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writePhylipFile</h4>
<pre>public&nbsp;void&nbsp;writePhylipFile(java.lang.String&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>fullyPathQualifiedFileName</code> - - name of the whole</dd>
<dd><code>useOriginalTaxonNames</code> - - or the truncated sequence names hash (first 6 chars of original name + underscore + 3-digit padded UID)</dd>
</dl>
</li>
</ul>
<a name="renameTaxon-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>renameTaxon</h4>
<pre>public&nbsp;boolean&nbsp;renameTaxon(java.lang.String&nbsp;oldName,
                           java.lang.String&nbsp;newName)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>oldName</code> - - the old name of the taxon (existing name)</dd>
<dd><code>newName</code> - - the new name to change it to</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if successful.
 
 This is a VERY quickly written and tested name-changing routine.
 TODO could do with a LOT more testing... 31/10/2011</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>31/10/2011</dd>
</dl>
</li>
</ul>
<a name="getFullSitesLnL-java.util.TreeMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getFullSitesLnL</h4>
<pre>public&nbsp;<a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</a>&nbsp;getFullSitesLnL(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>SitePatternsSSLS</code> - - a TreeMap<String, Float> of site patterns and their lnL values</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>fullSSLS - the sitewise lnL for each site in the alignment.</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>2011/11/30</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats"><code>DataSeries</code></a></dd>
</dl>
</li>
</ul>
<a name="clone--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clone</h4>
<pre>public&nbsp;<a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">AlignedSequenceRepresentation</a>&nbsp;clone()</pre>
<dl>
<dt><span class="overrideSpecifyLabel">Overrides:</span></dt>
<dd><code>clone</code>&nbsp;in class&nbsp;<code>java.lang.Object</code></dd>
</dl>
</li>
</ul>
<a name="filterForMissingData-int-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>filterForMissingData</h4>
<pre>public&nbsp;void&nbsp;filterForMissingData(int&nbsp;filter,
                                 boolean&nbsp;filterByFactor)
                          throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/FilterOutOfAllowableRangeException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">FilterOutOfAllowableRangeException</a></pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>filter</code> - - either the number of missing taxa at any given site, or the proportion (0-100 out of 100) of taxa that are allowed to have missing data at any given site [e.g. if the ratio (gapped taxa/total taxa) > (filter/100), for any given site]</dd>
<dd><code>filterByFactor</code> - - false to just use number of taxa missing as the threshold, true to use proportions.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/FilterOutOfAllowableRangeException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">FilterOutOfAllowableRangeException</a></code> - The allowable filter bounds are:
        0<filter�100  - when using filterFactor = true;
  0<numberOfTaxa        - when using filterFactor = false (e.g., threshold no. of gaps per site.</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>0.0.1 r63/4</dd>
</dl>
</li>
</ul>
<a name="getSequenceChars-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequenceChars</h4>
<pre>public&nbsp;char[]&nbsp;getSequenceChars(int&nbsp;sequenceIndex)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>-</code> - sequenceIndex - which sequence to get (not guaranteed sort order)</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>- a char array of sequence</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>- 09/08/2012</dd>
</dl>
</li>
</ul>
<a name="getTaxon-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTaxon</h4>
<pre>public&nbsp;java.lang.String&nbsp;getTaxon(int&nbsp;taxonIndex)</pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>-</code> - sequenceIndex - which taxon to get (not guaranteed sort order)</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>- the taxon name (String)</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>- 09/08/2012</dd>
</dl>
</li>
</ul>
<a name="stripGapsPreserveCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>stripGapsPreserveCodons</h4>
<pre>public&nbsp;void&nbsp;stripGapsPreserveCodons()
                             throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>-</code> - null</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>-</code> - SequenceTypeNotSupportedException</dd>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code></dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>- 21/08/2012</dd>
</dl>
</li>
</ul>
<a name="hasWholeNumberOfCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>hasWholeNumberOfCodons</h4>
<pre>public&nbsp;boolean&nbsp;hasWholeNumberOfCodons()
                               throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Boolean hasWholeNumberOfCodons - indicates whether no.codons/3 is an integer (whole codons) or not (partial codons)</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code> - A method that checks whether a whole number of codons are present, e.g. if (this.numberOfSites/3) has a remainder (% > 0)
 This method should throw a SequenceTypeNotSupportedException if AA or UNDEFINED sequence types are used</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>21/08/2012</dd>
</dl>
</li>
</ul>
<a name="trimToWholeNumberOfCodons--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>trimToWholeNumberOfCodons</h4>
<pre>public&nbsp;void&nbsp;trimToWholeNumberOfCodons()
                               throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<dl>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code></dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>21/08/2012</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#hasWholeNumberOfCodons--"><code>hasWholeNumberOfCodons()</code></a></dd>
</dl>
</li>
</ul>
<a name="engapStopCodonsInDNA--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>engapStopCodonsInDNA</h4>
<pre>public&nbsp;void&nbsp;engapStopCodonsInDNA()
                          throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<dl>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code></dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>21/08/2012</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><code>hasWholeNumberOfCodons(), trimToWholeNumberOfCodons, stripGapsPreserveCodons.</code></dd>
</dl>
</li>
</ul>
<a name="getInputIndices--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getInputIndices</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.Integer,java.lang.Integer&gt;&nbsp;getInputIndices()</pre>
</li>
</ul>
<a name="getAminoAcidsAsFitchStates--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAminoAcidsAsFitchStates</h4>
<pre>public&nbsp;java.util.HashMap&lt;java.lang.String,java.util.HashSet&lt;java.lang.String&gt;[]&gt;&nbsp;getAminoAcidsAsFitchStates()</pre>
<div class="block">A method to return a HashMap<String,HashSet<String>[]> of taxon names and sequences as states for a Fitch parsimony reconstruction by uk.ac.qmul.sbcs.evolution.convergence.TreeNode</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>Taxon names and sequences as HashSet<String>[]</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>2013/05/23</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TreeNode.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><code>TreeNode</code></a></dd>
</dl>
</li>
</ul>
<a name="calculateAlignmentStats-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>calculateAlignmentStats</h4>
<pre>public&nbsp;void&nbsp;calculateAlignmentStats(boolean&nbsp;evaluateGaps)</pre>
<div class="block">A utility method to populate basic stats (mean sitewise entropy; mean taxonwise longest ungapped sequence) for reporting.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>evaluateGaps</code> - : are gaps to be evaluated in the entropy calculations?</dd>
<dt><span class="simpleTagLabel">Since:</span></dt>
<dd>r188, 2013/07/29</dd>
</dl>
</li>
</ul>
<a name="isAA--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isAA</h4>
<pre>public&nbsp;boolean&nbsp;isAA()</pre>
<div class="block">Whether this alignment has been translated or not</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>- TRUE if alignment's SequenceCodingType is 'AA'</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence"><code>SequenceCodingType</code></a></dd>
</dl>
</li>
</ul>
<a name="loadSequences-java.util.ArrayList-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>loadSequences</h4>
<pre>public&nbsp;boolean&nbsp;loadSequences(java.util.ArrayList&lt;java.lang.String&gt;&nbsp;inputData,
                             boolean&nbsp;reportInputRead)
                      throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</a></pre>
<dl>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</a></code></dd>
</dl>
</li>
</ul>
<a name="printSharedSubs--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printSharedSubs</h4>
<pre>public&nbsp;void&nbsp;printSharedSubs()</pre>
</li>
</ul>
<a name="printSharedPrivateSubs--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printSharedPrivateSubs</h4>
<pre>public&nbsp;void&nbsp;printSharedPrivateSubs()</pre>
</li>
</ul>
<a name="printSharedPrivateSubs-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>printSharedPrivateSubs</h4>
<pre>public&nbsp;void&nbsp;printSharedPrivateSubs(java.lang.String&nbsp;taxon_1,
                                   java.lang.String&nbsp;taxon_2)</pre>
<div class="block">Prints substitutions shared between two taxa exclusively (no other taxa) to stdout</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>taxon_1</code> - </dd>
<dd><code>taxon_2</code> - </dd>
</dl>
</li>
</ul>
<a name="simulateConvergenceInInvariantSites-java.lang.String-java.lang.String:A-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>simulateConvergenceInInvariantSites</h4>
<pre>public&nbsp;void&nbsp;simulateConvergenceInInvariantSites(java.lang.String&nbsp;masterTaxon,
                                                java.lang.String[]&nbsp;taxaToForce,
                                                int&nbsp;numberOfSitesToConverge)
                                         throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</a>,
                                                <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</a></pre>
<div class="block">Purpose:        constrain a subset of taxa to have the same AA
        <p>execution:   
 <pre>for i in num sites{
        random indices
        get target master taxon as char array /// NB: codon triplets
        read array[indices]
        replace
}</pre></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>masterTaxon</code> - - The taxon that is being treated as the master. Sites will be converged to this taxon's sequence randomly</dd>
<dd><code>taxaToForce</code> - - The taxa that in which convergence is to be simulated.</dd>
<dd><code>numberOfSitesToConverge</code> - - The number of sites to be forced to converge.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</a></code> - - In cases where the taxaToForce are not present in the alignment (check case of taxa?)</dd>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</a></code></dd>
</dl>
</li>
</ul>
<a name="forceTranslationIgnoringSequenceType--">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>forceTranslationIgnoringSequenceType</h4>
<pre>@Deprecated
public&nbsp;void&nbsp;forceTranslationIgnoringSequenceType()
                                                      throws <a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></pre>
<div class="block"><span class="deprecatedLabel">Deprecated.</span>&nbsp;<span class="deprecationComment">Debug only.</span></div>
<div class="block">Forces the alignment to be translated ignoring which sequence type has been assumed. Debug only.</div>
<dl>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceTypeNotSupportedException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">SequenceTypeNotSupportedException</a></code></dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/AlignedSequenceRepresentation.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignmentTranslationTableType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top">Frames</a></li>
<li><a href="AlignedSequenceRepresentation.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
