<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_29) on Fri Jan 06 15:00:22 GMT 2012 -->
<TITLE>
AlignedSequenceRepresentation
</TITLE>

<META NAME="date" CONTENT="2012-01-06">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="AlignedSequenceRepresentation";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlignedSequenceRepresentation.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/NewickTreeRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlignedSequenceRepresentation.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
uk.ac.qmul.sbcs.evolution.convergence</FONT>
<BR>
Class AlignedSequenceRepresentation</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../../resources/inherit.gif" ALT="extended by "><B>uk.ac.qmul.sbcs.evolution.convergence.AlignedSequenceRepresentation</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>AlignedSequenceRepresentation</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>09/19/2011</DD>
<DT><B>Version:</B></DT>
  <DD>0.1</DD>
<DT><B>Author:</B></DT>
  <DD>Joe Parker, <a href="http://code.google.com/a/eclipselabs.org/u/joeparkerandthemegahairymen/">Kitson Consulting Ltd / Queen Mary, University of London.</a></DD>
</DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#AlignedSequenceRepresentation()">AlignedSequenceRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputFileDatatype()">determineInputFileDatatype</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInputSequenceType()">determineInputSequenceType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#determineInvariantSites()">determineInvariantSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getFullSitesLnL(java.util.TreeMap)">getFullSitesLnL</A></B>(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfSites()">getNumberOfSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getNumberOfTaxa()">getNumberOfTaxa</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.ArrayList&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getRawInput()">getRawInput</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeSet&lt;java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTaxaList()">getTaxaList</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTransposedSites()">getTransposedSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;java.lang.String,java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#getTruncatedNamesHash()">getTruncatedNamesHash</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#loadSequences(java.io.File, boolean)">loadSequences</A></B>(java.io.File&nbsp;inputFile,
              boolean&nbsp;reportInputRead)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#PhymlSequenceRespresentation()">PhymlSequenceRespresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfInvariantSites()">printNumberOfInvariantSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfSites()">printNumberOfSites</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printNumberOfTaxa()">printNumberOfTaxa</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#printShortSequences(int)">printShortSequences</A></B>(int&nbsp;printLimit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readFastaFile()">readFastaFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readNexusFile()">readNexusFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readPhylipFile()">readPhylipFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#readXMLFile()">readXMLFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeStopCodons()">removeStopCodons</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeTaxa(java.lang.String[])">removeTaxa</A></B>(java.lang.String[]&nbsp;taxaToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#removeUnambiguousGaps()">removeUnambiguousGaps</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#renameTaxon(java.lang.String, java.lang.String)">renameTaxon</A></B>(java.lang.String&nbsp;oldName,
            java.lang.String&nbsp;newName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#simulateConvergence(java.lang.String, java.lang.String[], int)">simulateConvergence</A></B>(java.lang.String&nbsp;masterTaxon,
                    java.lang.String[]&nbsp;taxaToForce,
                    int&nbsp;numberOfSitesToConverge)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;purpose:        constrain a subset of taxa to have the same AA
        execution:      for i in num sites
                                        random indices
                                        get target master taxon as char array /// NB: codon triplets
                                        read array[indices]
                                        replace</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#translate(boolean)">translate</A></B>(boolean&nbsp;suppressErrors)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeFastaFile()">writeFastaFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeNexusFile()">writeNexusFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.io.File, boolean)">writePhylipFile</A></B>(java.io.File&nbsp;fullyPathQualifiedFileName,
                boolean&nbsp;useOriginalTaxonNames)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.lang.String)">writePhylipFile</A></B>(java.lang.String&nbsp;fullyPathQualifiedFileName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writePhylipFile(java.lang.String, boolean)">writePhylipFile</A></B>(java.lang.String&nbsp;fullyPathQualifiedFileName,
                boolean&nbsp;useOriginalTaxonNames)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html#writeXMLFile()">writeXMLFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="AlignedSequenceRepresentation()"><!-- --></A><H3>
AlignedSequenceRepresentation</H3>
<PRE>
public <B>AlignedSequenceRepresentation</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="PhymlSequenceRespresentation()"><!-- --></A><H3>
PhymlSequenceRespresentation</H3>
<PRE>
public void <B>PhymlSequenceRespresentation</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="loadSequences(java.io.File, boolean)"><!-- --></A><H3>
loadSequences</H3>
<PRE>
public void <B>loadSequences</B>(java.io.File&nbsp;inputFile,
                          boolean&nbsp;reportInputRead)
                   throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></PRE>
<DL>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/TaxaLimitException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util">TaxaLimitException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="determineInputFileDatatype()"><!-- --></A><H3>
determineInputFileDatatype</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceFileFormat.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceFileFormat</A> <B>determineInputFileDatatype</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineInputSequenceType()"><!-- --></A><H3>
determineInputSequenceType</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/SequenceCodingType.html" title="enum in uk.ac.qmul.sbcs.evolution.convergence">SequenceCodingType</A> <B>determineInputSequenceType</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="readPhylipFile()"><!-- --></A><H3>
readPhylipFile</H3>
<PRE>
public void <B>readPhylipFile</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>31/10/2011
 
 Added methods to parse lines that include spaces within sequence lines.
 NOTE that this assumes rawInput has NO blank lines (before data, at least).</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaFile()"><!-- --></A><H3>
readFastaFile</H3>
<PRE>
public void <B>readFastaFile</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>31/10/2011
 RESOLVED: 31/10/2011: TODO testing revealing some very strange behaviour in some fasta reads. work out why.
 A number of errors - 
                lastname and lastdata initialised wrongly to null not ""
                taxon names added as substr(1) not substr(1,length-1) so not now truncated
                (incidental) loadSequences() now takes boolean arg for line reporting
                MAJOR error:    lastname was always being reinitialised after sequenceHash.add() to "", so half of the time
                                                e.g. some taxa names are ignored, and their data entered in the other taxa, doubling line length
                MAJOR error:    the final taxon was never being added to sequenceHash as its match condition was not being filled.</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="readNexusFile()"><!-- --></A><H3>
readNexusFile</H3>
<PRE>
public void <B>readNexusFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="readXMLFile()"><!-- --></A><H3>
readXMLFile</H3>
<PRE>
public void <B>readXMLFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.lang.String)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.lang.String&nbsp;fullyPathQualifiedFileName)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.io.File, boolean)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.io.File&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeFastaFile()"><!-- --></A><H3>
writeFastaFile</H3>
<PRE>
public void <B>writeFastaFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeNexusFile()"><!-- --></A><H3>
writeNexusFile</H3>
<PRE>
public void <B>writeNexusFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeXMLFile()"><!-- --></A><H3>
writeXMLFile</H3>
<PRE>
public void <B>writeXMLFile</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="simulateConvergence(java.lang.String, java.lang.String[], int)"><!-- --></A><H3>
simulateConvergence</H3>
<PRE>
public void <B>simulateConvergence</B>(java.lang.String&nbsp;masterTaxon,
                                java.lang.String[]&nbsp;taxaToForce,
                                int&nbsp;numberOfSitesToConverge)
                         throws <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A>,
                                <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></PRE>
<DL>
<DD>purpose:        constrain a subset of taxa to have the same AA
        execution:      for i in num sites
                                        random indices
                                        get target master taxon as char array /// NB: codon triplets
                                        read array[indices]
                                        replace
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>masterTaxon</CODE> - - The taxon that is being treated as the master. Sites will be converged to this taxon's sequence randomly<DD><CODE>taxaToForce</CODE> - - The taxa that in which convergence is to be simulated.<DD><CODE>numberOfSitesToConverge</CODE> - - The number of sites to be forced to converge.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/TaxaAbsentFromAlignmentException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">TaxaAbsentFromAlignmentException</A></CODE> - - In cases where the taxaToForce are not present in the alignment (check case of taxa?)
<DD><CODE><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/VariantSitesUnavailableException.html" title="class in uk.ac.qmul.sbcs.evolution.convergence">VariantSitesUnavailableException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="printShortSequences(int)"><!-- --></A><H3>
printShortSequences</H3>
<PRE>
public void <B>printShortSequences</B>(int&nbsp;printLimit)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="translate(boolean)"><!-- --></A><H3>
translate</H3>
<PRE>
public void <B>translate</B>(boolean&nbsp;suppressErrors)
               throws java.lang.Exception</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>suppressErrors</CODE> - - controls verbose stdout commentary (TRUE = less verbose)
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE><DT><B>Since:</B></DT>
  <DD>05/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeTaxa(java.lang.String[])"><!-- --></A><H3>
removeTaxa</H3>
<PRE>
public void <B>removeTaxa</B>(java.lang.String[]&nbsp;taxaToRemove)</PRE>
<DL>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>06/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeUnambiguousGaps()"><!-- --></A><H3>
removeUnambiguousGaps</H3>
<PRE>
public void <B>removeUnambiguousGaps</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>06/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="removeStopCodons()"><!-- --></A><H3>
removeStopCodons</H3>
<PRE>
public void <B>removeStopCodons</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineInvariantSites()"><!-- --></A><H3>
determineInvariantSites</H3>
<PRE>
public void <B>determineInvariantSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getRawInput()"><!-- --></A><H3>
getRawInput</H3>
<PRE>
public java.util.ArrayList&lt;java.lang.String&gt; <B>getRawInput</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfSites()"><!-- --></A><H3>
getNumberOfSites</H3>
<PRE>
public int <B>getNumberOfSites</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>the maximum numberOfSites in the data alignment</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfTaxa()"><!-- --></A><H3>
getNumberOfTaxa</H3>
<PRE>
public int <B>getNumberOfTaxa</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>the numberOfTaxa</DL>
</DD>
</DL>
<HR>

<A NAME="getTaxaList()"><!-- --></A><H3>
getTaxaList</H3>
<PRE>
public java.util.TreeSet&lt;java.lang.String&gt; <B>getTaxaList</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>the taxaList (String ArrayList of untruncated taxa names.)</DL>
</DD>
</DL>
<HR>

<A NAME="getTruncatedNamesHash()"><!-- --></A><H3>
getTruncatedNamesHash</H3>
<PRE>
public java.util.HashMap&lt;java.lang.String,java.lang.String&gt; <B>getTruncatedNamesHash</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>the truncatedNamesHash</DL>
</DD>
</DL>
<HR>

<A NAME="getTransposedSites()"><!-- --></A><H3>
getTransposedSites</H3>
<PRE>
public java.lang.String[] <B>getTransposedSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfSites()"><!-- --></A><H3>
printNumberOfSites</H3>
<PRE>
public void <B>printNumberOfSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfTaxa()"><!-- --></A><H3>
printNumberOfTaxa</H3>
<PRE>
public void <B>printNumberOfTaxa</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printNumberOfInvariantSites()"><!-- --></A><H3>
printNumberOfInvariantSites</H3>
<PRE>
public void <B>printNumberOfInvariantSites</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writePhylipFile(java.lang.String, boolean)"><!-- --></A><H3>
writePhylipFile</H3>
<PRE>
public void <B>writePhylipFile</B>(java.lang.String&nbsp;fullyPathQualifiedFileName,
                            boolean&nbsp;useOriginalTaxonNames)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="renameTaxon(java.lang.String, java.lang.String)"><!-- --></A><H3>
renameTaxon</H3>
<PRE>
public boolean <B>renameTaxon</B>(java.lang.String&nbsp;oldName,
                           java.lang.String&nbsp;newName)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>oldName</CODE> - - the old name of the taxon (existing name)<DD><CODE>newName</CODE> - - the new name to change it to
<DT><B>Returns:</B><DD>true if successful.
 
 This is a VERY quickly written and tested name-changing routine.
 TODO could do with a LOT more testing... 31/10/2011<DT><B>Since:</B></DT>
  <DD>31/10/2011</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getFullSitesLnL(java.util.TreeMap)"><!-- --></A><H3>
getFullSitesLnL</H3>
<PRE>
public <A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats">DataSeries</A> <B>getFullSitesLnL</B>(java.util.TreeMap&lt;java.lang.String,java.lang.Float&gt;&nbsp;SitePatternsSSLS)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>SitePatternsSSLS</CODE> - - a TreeMap<String, Float> of site patterns and their lnL values
<DT><B>Returns:</B><DD>fullSSLS - the sitewise lnL for each site in the alignment.<DT><B>Since:</B></DT>
  <DD>2011/11/30</DD>
<DT><B>See Also:</B><DD><A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/util/stats/DataSeries.html" title="class in uk.ac.qmul.sbcs.evolution.convergence.util.stats"><CODE>DataSeries</CODE></A></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlignedSequenceRepresentation.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../../../uk/ac/qmul/sbcs/evolution/convergence/NewickTreeRepresentation.html" title="class in uk.ac.qmul.sbcs.evolution.convergence"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/ac/qmul/sbcs/evolution/convergence/AlignedSequenceRepresentation.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlignedSequenceRepresentation.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
